### PR TITLE
fix: grant role before creating database

### PR DESCRIPTION
**Pull request by ltalirz**
In some postgresql installations, the following fails:

	template1=> create user test;
	CREATE ROLE
	template1=> create database test owner test;
	ERROR:  must be member of role "test"

This is fixed by

	template1=> grant test to azureuser;
	GRANT ROLE
	template1=> create database test owner test;
	CREATE DATABASE

Furthermore, granting privileges to the owner is superfluous according to the postgresql docs, so we drop it.

The investigation of this issue also surfaces a bug in the error message for "manual database creation", and that the `--su-db-name` was not passed on from the commandline. This is fixed in passing.

### CODE REVIEW

**Comment by ltalirz** [comment]
Ha, that is interesting.

Somehow, one of these fixes now surfaces the (expected) problem with deleting the test database.

I guess we'll need to do work item 1 from https://github.com/aiidateam/aiida-core/issues/3762#issuecomment-1318623586 before continuing

**Comment by ltalirz** [comment]
The error is

``` 
+ verdi quicksetup --non-interactive --profile default --email aiida@localhost --first-name Giuseppe --last-name Verdi --institution Khedivial --db-backend core.psql_dos
Error: Oops! quicksetup was unable to create the AiiDA database for you.
See `verdi quicksetup -h` for how to specify non-standard parameters for the postgresql connection.
Alternatively, create the AiiDA database yourself: 
Run the following commands as a UNIX user with access to PostgreSQL (Ubuntu: $ sudo su postgres):

	$ psql template1
	==> CREATE USER "aiida_qs_aiida_477d3dfc78a2042156110cb00ae[36](https://github.com/aiidateam/aiida-core/actions/runs/3644014214/jobs/6152813665#step:7:37)18f" WITH PASSWORD '<password>' 
	==> GRANT "aiida_qs_aiida_477d3dfc78a2042156110cb00ae3618f" TO "postgres"
	==> CREATE DATABASE "default_aiida_477d3dfc78a2042156110cb00ae3618f" OWNER "aiida_qs_aiida_477d3dfc78a2042156110cb00ae3618f" ENCODING 'UTF8' LC_COLLATE='en_US.UTF-8' LC_CTYPE='en_US.UTF-8' TEMPLATE=template0

and then use `verdi setup` instead

Traceback (most recent call last):
  File "/opt/conda/bin/verdi", line 8, in <module>
    sys.exit(verdi())
  File "/opt/conda/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/opt/conda/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/opt/conda/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/lib/python3.8/site-packages/click/core.py", line 1[40](https://github.com/aiidateam/aiida-core/actions/runs/3644014214/jobs/6152813665#step:7:41)4, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/conda/lib/python3.8/site-packages/click/core.py", line 7[60](https://github.com/aiidateam/aiida-core/actions/runs/3644014214/jobs/6152813665#step:7:61), in invoke
    return __callback(*args, **kwargs)
  File "/opt/conda/lib/python3.8/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/opt/conda/lib/python3.8/site-packages/aiida/cmdline/commands/cmd_setup.py", line 185, in quicksetup
    raise exception
  File "/opt/conda/lib/python3.8/site-packages/aiida/cmdline/commands/cmd_setup.py", line 174, in quicksetup
    db_username, db_name = postgres.create_dbuser_db_safe(dbname=db_name, dbuser=db_username, dbpass=db_password)
  File "/opt/conda/lib/python3.8/site-packages/aiida/manage/external/postgres.py", line 194, in create_dbuser_db_safe
    self.create_dbuser(dbuser=dbuser, dbpass=dbpass)
  File "/opt/conda/lib/python3.8/site-packages/aiida/manage/external/postgres.py", line 112, in create_dbuser
    self.execute(_GRANT_ROLE_COMMAND.format(dbuser, self.dsn['user']))
  File "/opt/conda/lib/python3.8/site-packages/pgsu/__init__.py", line 120, in execute
    return _execute_psyco(command, dsn)
  File "/opt/conda/lib/python3.8/site-packages/pgsu/__init__.py", line 2[61](https://github.com/aiidateam/aiida-core/actions/runs/3644014214/jobs/6152813665#step:7:62), in _execute_psyco
    cursor.execute(command)
psycopg2.errors.UndefinedObject: role "None" does not exist
``` 

I.e. somehow `self.dsn['user']` is `None`, while `postgres` is correctly passed to the constructor (and the "manual instructions" look correct).

**Comment by ltalirz** [comment]
I see - inside the container, going via the default `postgres` user doesn't work, and AiiDA ends up using peer authentication (which PGSU represents internally as user `None`).

We can [use the `current_user` shorthand in postgresql](https://www.postgresql.org/docs/12/functions-info.html) when granting the role.

File: aiida/cmdline/commands/cmd_setup.py [hunk 1]
```diff
@@ -177,8 +178,8 @@ def quicksetup(
                 'Oops! quicksetup was unable to create the AiiDA database for you.',
                 'See `verdi quicksetup -h` for how to specify non-standard parameters for the postgresql connection.\n'
                 'Alternatively, create the AiiDA database yourself: ',
-                manual_setup_instructions(dbuser=su_db_username,
-                                          dbname=su_db_name), '', 'and then use `verdi setup` instead', ''
+                manual_setup_instructions(db_username=db_username,
```
Comment (sphuber) [inline aiida/cmdline/commands/cmd_setup.py:181]: Not concerning this line, but line 184: should we just reraise the exception? Isn't it better to simply use `echo.echo_critical`? Usually showing a scary traceback by default to a user from a CLI command is not a good idea.

Comment (sphuber) [reply aiida/cmdline/commands/cmd_setup.py:181]: Also just noted that the instructions above (to use `verdi quicksetup -h` for help) were wrong for a long time and only (coincidentally) got fixed just two weeks ago in 6469e2360680ff309c2275dbe205b979e2987ac6  :smile:

File: aiida/manage/external/postgres.py [hunk 2]
```diff
@@ -108,6 +108,8 @@ def create_dbuser(self, dbuser, dbpass, privileges=''):
             self.connection_mode == PostgresConnectionMode.PSYCOPG
         """
         self.execute(_CREATE_USER_COMMAND.format(dbuser, dbpass, privileges))
+        # this is needed on some postgresql installations
```
Comment (sphuber) [inline aiida/manage/external/postgres.py:111]: Could you please add more info on what this is doing? It seems quite weird to `GRANT` a user to `current_user`. What does that mean? Grant all rights that `dbuser` has also to `current_user`?

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @ltalirz some minor comments

File: aiida/manage/external/postgres.py (hunk 2 above)
Comment (ltalirz) [inline aiida/manage/external/postgres.py:111]:
```suggestion
        # Ensure the database superuser (current_user) has the rights to grant `dbuser` access to new databases.
        # Required for some postgresql installations.
```

File: aiida/cmdline/commands/cmd_setup.py (hunk 1 above)
Comment (ltalirz) [reply aiida/cmdline/commands/cmd_setup.py:181]: Sometimes the traceback can also help to see where the problem is... let me know if you want me to change it

File: aiida/manage/external/postgres.py (hunk 2 above)
Comment (ltalirz) [reply aiida/manage/external/postgres.py]: I've updated the comment. Let me know whether it is understandable now

File: aiida/manage/external/postgres.py [hunk 3]
```diff
@@ -108,6 +108,9 @@ def create_dbuser(self, dbuser, dbpass, privileges=''):
             self.connection_mode == PostgresConnectionMode.PSYCOPG
         """
         self.execute(_CREATE_USER_COMMAND.format(dbuser, dbpass, privileges))
+        # Ensure the database superuser (current_user) has the rights to grant `dbuser` access to new databases.
```
Comment (ltalirz) [inline aiida/manage/external/postgres.py:111]:
```suggestion
        # Ensure the database superuser (current_user) has the right to make `dbuser` the owner of new databases.
```
```suggestion
        # Ensure the database superuser (current_user) has the rights to grant `dbuser` access to new databases.
```
```suggestion
        # Ensure the database superuser (current_user) has the rights to grant `dbuser` access to new databases.
```

File: aiida/cmdline/commands/cmd_setup.py (hunk 1 above)
Comment (sphuber) [reply aiida/cmdline/commands/cmd_setup.py:181]: There is a reusable option that we have `--traceback` that will optionally show the traceback which I think can be useful. But we can leave it for now, it is not critical

File: aiida/manage/external/postgres.py (hunk 2 above)
Comment (sphuber) [reply aiida/manage/external/postgres.py]: Cheers.

**Review by sphuber — APPROVED** [review]
Thanks @ltalirz

**Comment by sphuber** [comment]
@ltalirz did this address #5827 ? Think so right?

**Comment by ltalirz** [comment]
Yes, that is fixed now (meaning you can now use AiiDA with the hosted Azure Database for Postgresql

**Pull request merged**